File: IT/Multithreading/Esercizi/Java/Esercizio15.java
/* file Esercizio15.java */

import java.util.concurrent.atomic.AtomicInteger;

class TargetAdd implements Runnable {
	private Object monitor;
	private AtomicInteger value;
 
	public TargetAdd(Object monitor, AtomicInteger value) {
		this.monitor = monitor;
		this.value = value;
	}

	@Override
	public void run() {
		synchronized(monitor) {
			String name = Thread.currentThread().getName();
			while (value.get() > 0) {
				try {
					monitor.wait();
				}
				catch (InterruptedException e) {
					e.printStackTrace();
				}
			}

			System.out.println(name + " - variabile (prima dell'incremento): " + value);
			int newValue = value.incrementAndGet();
			System.out.println(name + " - variabile (dopo l'incremento): " + newValue);
			monitor.notify();
		}
	}
}

class TargetSubtr implements Runnable {
	private Object monitor;
	private AtomicInteger value;
 
	public TargetSubtr(Object monitor, AtomicInteger value) {
		this.monitor = monitor;
		this.value = value;
	}

	@Override
	public void run() {
		synchronized(monitor) {
			String name = Thread.currentThread().getName();
			while (value.get() == 0) {
				try {
					monitor.wait();
				}
				catch (InterruptedException e) {
					e.printStackTrace();
				}
			}

			System.out.println(name + " - variabile (prima del decremento): " + value);
			int newValue = value.decrementAndGet();
			System.out.println(name + " - variabile (dopo il decremento): " + newValue);
			monitor.notify();
		}
	}
}

public class Esercizio15 {
	public static void main(String [] args) {
		final Object monitor = new Object();
		AtomicInteger value = new AtomicInteger(0);

		Thread t1 = new Thread(new TargetAdd(monitor, value));
		Thread t2 = new Thread(new TargetSubtr(monitor, value));

		t1.start();
		t2.start();
	}
}

File: IT/Multithreading/Esercizi/C++/esercizio15.cpp
/*
file esercizio15.cpp
compilare con: g++ -std=c++11 esercizio15.cpp -lpthread
*/

#include <iostream>
#include <thread>
#include <condition_variable>

using std::ref;
using std::thread;
using std::mutex;
using std::unique_lock;
using std::condition_variable;

void add(int& value, mutex& _mutex, condition_variable& condition, int i)
{
	unique_lock<mutex> lock(_mutex);

	while (value > 0) condition.wait(lock);

	printf("thread %d - variabile (prima dell'incremento): %d\n", i, value/*.load()*/);

	value++;

	printf("thread %d - variabile (dopo l'incremento): %d\n", i, value/*.load()*/);

	condition.notify_one();
}

void subtr(int& value, mutex& _mutex, condition_variable& condition, int i)
{
	unique_lock<mutex> lock(_mutex);

	if (value == 0) condition.wait(lock);

	printf("thread %d - variabile (prima del decremento): %d\n", i, value);

	value--;

	printf("thread %d - variabile (dopo il decremento): %d\n", i, value);

	condition.notify_one();
}

int main()
{
	int value = 0;
	mutex lock;
	condition_variable condition;

	thread t1(add, ref(value), ref(lock), ref(condition), 1);
	thread t2(subtr, ref(value), ref(lock), ref(condition), 2);

	t1.join();
	t2.join();
}

